### PR TITLE
Refactored reputation w/ sampling

### PR DESCRIPTION
**Pull request by antonypap**
Logic should be clear now, and no more pesky bugs :)

### CODE REVIEW

File: pkg/infra/teams/team3/election_strat.go [hunk 1]
```diff
@@ -184,81 +158,49 @@ func (a *AgentThree) calcW2(baseAgent agent.BaseAgent, w2 float64) float64 {
...
-	}
 	productivity := 5.0
 	needs := 5.0
 	fairness := make(map[commons.ID]float64)
-	sortedfairness := make(map[commons.ID]float64)
 
-	itr := agentState.Iterator()
-	counter1 := 1
+	// Number of agents to sample for KA (fixed)
+	sampleLength := int(math.Max(float64(a.numAgents)*a.sample_percent, float64(vAS.Len())))
```
Comment (MattSScott) [inline pkg/infra/teams/team3/election_strat.go]: Extract right hand side into variables (for checking)

File: pkg/infra/teams/team3/election_strat.go [hunk 2]
```diff
@@ -184,81 +158,49 @@ func (a *AgentThree) calcW2(baseAgent agent.BaseAgent, w2 float64) float64 {
...
+	cnt := 0
 
-	for i := 1; i <= int(math.Ceil(float64(agentState.Len())*a.sample_percent)); {
-		if counter1%sample == 0 {
-			id, hiddenState, _ := itr.Next()
+	// Use randomness of maps in go to take n random samples
+	for _, id := range ids {
+		if cnt == sampleLength {
+			break
```
Comment (MattSScott) [inline pkg/infra/teams/team3/election_strat.go]: You can just return `fairness` here, I believe

**Review by MattSScott — COMMENTED** [review]
Looks good - minor recommendations

File: pkg/infra/teams/team3/election_strat.go (hunk 1 above)
Comment (BusterBlack) [reply pkg/infra/teams/team3/election_strat.go]: also, should we be using the Max function here? I think it should be Min()

File: pkg/infra/teams/team3/election_strat.go [hunk 3]
```diff
@@ -184,81 +158,49 @@ func (a *AgentThree) calcW2(baseAgent agent.BaseAgent, w2 float64) float64 {
...
+		} else {
+			// Init values on 1st lvl
+			if view.CurrentLevel() == 1 {
```
Comment (BusterBlack) [inline pkg/infra/teams/team3/election_strat.go]: some agents might not be initialised on lv.1... check if ID exists in fairness map, and add if it doesn't with init values.

**Comment by BusterBlack** [comment]
LGTM

**Review by BusterBlack — APPROVED** [review]

**Pull request merged**